Source/Diagnostics/ReducedDiags/FieldMomentum.cpp

```diff
@@ -0,0 +1,217 @@
+    {
+        // Get MultiFab data at given refinement level
+        const amrex::MultiFab & Ex = warpx.getEfield(lev, 0);
+        const amrex::MultiFab & Ey = warpx.getEfield(lev, 1);
+        const amrex::MultiFab & Ez = warpx.getEfield(lev, 2);
+        const amrex::MultiFab & Bx = warpx.getBfield(lev, 0);
+        const amrex::MultiFab & By = warpx.getBfield(lev, 1);
+        const amrex::MultiFab & Bz = warpx.getBfield(lev, 2);
+
+        // Cell-centered index type
+        const amrex::GpuArray<int,3> cc{0,0,0};
+        // Coarsening ratio (no coarsening)
+        const amrex::GpuArray<int,3> cr{1,1,1};
+        // Reduction component (fourth component in Array4)
+        constexpr int comp = 0;
+
+        // Index type (staggering) of each MultiFab
+        // (with third component set to zero in 2D)
+        amrex::GpuArray<int,3> Ex_stag{0,0,0};
+        amrex::GpuArray<int,3> Ey_stag{0,0,0};
+        amrex::GpuArray<int,3> Ez_stag{0,0,0};
+        amrex::GpuArray<int,3> Bx_stag{0,0,0};
+        amrex::GpuArray<int,3> By_stag{0,0,0};
+        amrex::GpuArray<int,3> Bz_stag{0,0,0};
+        for (int i = 0; i < AMREX_SPACEDIM; ++i)
+        {
+            Ex_stag[i] = Ex.ixType()[i];
+            Ey_stag[i] = Ey.ixType()[i];
+            Ez_stag[i] = Ez.ixType()[i];
+            Bx_stag[i] = Bx.ixType()[i];
+            By_stag[i] = By.ixType()[i];
+            Bz_stag[i] = Bz.ixType()[i];
+        }
+
+        amrex::ReduceOps<amrex::ReduceOpSum> reduce_op_ExB_x;
+        amrex::ReduceOps<amrex::ReduceOpSum> reduce_op_ExB_y;
+        amrex::ReduceOps<amrex::ReduceOpSum> reduce_op_ExB_z;
+
+        amrex::ReduceData<amrex::Real> reduce_data_ExB_x(reduce_op_ExB_x);
+        amrex::ReduceData<amrex::Real> reduce_data_ExB_y(reduce_op_ExB_y);
+        amrex::ReduceData<amrex::Real> reduce_data_ExB_z(reduce_op_ExB_z);
+
+        // Reduce operation type is the same (amrex::ReduceOpSum) for all components of (E x B)
+        using ReduceTuple = typename decltype(reduce_data_ExB_x)::Type;
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        // Loop over boxes, interpolate E,B data to cell centers
+        // and compute sum over cells of (E x B) components
+        for (amrex::MFIter mfi(Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+        {
+            const amrex::Box & box = enclosedCells(mfi.nodaltilebox());
+            const amrex::Array4<const amrex::Real> & Ex_arr = Ex[mfi].array();
+            const amrex::Array4<const amrex::Real> & Ey_arr = Ey[mfi].array();
+            const amrex::Array4<const amrex::Real> & Ez_arr = Ez[mfi].array();
+            const amrex::Array4<const amrex::Real> & Bx_arr = Bx[mfi].array();
+            const amrex::Array4<const amrex::Real> & By_arr = By[mfi].array();
+            const amrex::Array4<const amrex::Real> & Bz_arr = Bz[mfi].array();
+
+            // (E x B) along x
+            reduce_op_ExB_x.eval(box, reduce_data_ExB_x,
```

Comment:
@EZoni @atmyers just as a potential follow-up PR: do you think we can use a single reduce for the fields as well? That would allow us to read the field components twice and spawn only one kernel.

Reply:
Actually you already turn into a tuple, I guess that might already work?

Reply:
Good idea, I'll try to see if I can do something similar to what I did for particles. I'll mark the PR as [WIP] until I see if that works.

Reply:
Weiqun commented:
```C++
amrex::ReduceOps<ReduceOpSum, ReduceOpSum, ReduceOpSum> reduce_ops;
amrex::ReduceData<Real, Real, Real> reduce_data(reduce_ops);
for (MFIter ...) {
    reduce_ops.eval(box, reduce_data,
    [=] AMREX_GPU_DEVICE (int i, int j, int k) -> amrex::GpuTuple<Real, Real, Real>
    {
        const amrex::Real Ex_cc = ...;
        const amrex::Real Ey_cc = ...;
        const amrex::Real Ez_cc = ...;
        const amrex::Real Bx_cc = ...;
        const amrex::Real By_cc = ...;
        const amrex::Real Bz_cc = ...;
        return {Ey_cc * Bz_cc - Ez_cc * By_cc,
                Ez_cc * Bx_cc - Ex_cc * Bz_cc,
                Ex_cc * By_cc - Ey_cc * Bx_cc};
     });
}
auto r = reduce_data.value();
Real ExB_x = amrex::get<0>(r);
Real ExB_y = amrex::get<1>(r);
Real ExB_z = amrex::get<2>(r);
```

Reply:
Thank you, done in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/4224952b3e34d2607f48eae901c73881929fabaf.

Reply:
Hmm, looks like we could do this for the `FieldMaximum` (and potentially `RhoMaximum`) diag as well. I'll add this on my todo list.